src/eval/mod.rs

```diff
@@ -677,6 +644,64 @@ impl<R: ImportResolver> VirtualMachine<R> {
                         return Ok((RichTerm::new(Term::Fun(*x, t.clone()), pos), env));
                     }
                 }
+                // A match expression acts as a function (in Nickel, a match expression corresponds
+                // to the cases, and doesn't include the examined value).
+                //
+                // The behavior is the same as for a function: we look for an argument on the
+                // stack, and proceed to the evaluation of the match, or stop here otherwise. If
+                // found (let's call it `arg`), we evaluate `%match% arg cases default`, where
+                // `%match%` is the primitive operation `UnaryOp::Match` taking care of forcing the
+                // argument `arg` and doing the actual matching operation.
+                Term::Match { cases, default } => {
+                    if let Some((arg, pos_app)) = self.stack.pop_arg() {
+                        // Setting the stack to be as if we would have evaluated an application
+                        // `_ cases default`, where `default` is optional, and `_` is not relevant.
+
+                        let has_default = default.is_some();
+
+                        if let Some(t) = default {
+                            self.stack.push_arg(
+                                Closure {
+                                    body: t.clone(),
+                                    env: env.clone(),
+                                },
+                                pos,
+                            );
+                        }
+
+                        self.stack.push_arg(
+                            Closure {
+                                body: RichTerm::new(
+                                    Term::Record(RecordData::with_fields(cases.clone())),
+                                    pos,
+                                ),
+                                env: env.clone(),
+                            },
+                            pos,
+                        );
+
+                        // Now evaluating `%match% arg`, the left-most part of the application `%match%
+                        // arg cases default`, which is in fact a primop application.
+                        self.stack.push_op_cont(
+                            OperationCont::Op1(UnaryOp::Match { has_default }, pos_app),
```

Comment:
I think `has_default` here could just be `true`, right? As we've already destructed `default` above.

Reply:
Unless I'm misreading something, the `if let Some(t) = default` block ends line 670, so right here we may or may not have a default value.